Reduce bundle size, add code splitting

MarcelRobitaille opened the pull request:
I'm writing this as a PR only so you can see the code changes. I don't intend this to be merged as is. I am just trying to continue the code splitting discussion.

Fixes #455 
Fixes #1109 

I implemented "naive" code splitting. I say naive because I just did every different route. However, these routes are usually quite small. We might be better off including the truly tiny ones in the main bundle to minimize the number of network requests.

Here is the bundle analyzer for the `master` branch.
![image](https://user-images.githubusercontent.com/8503756/180922279-da292d7f-ee0a-41eb-bc91-ca3f137f6036.png)
- `cookbook-main.js` is 4.79 MB raw, 1.05 MB gzipped.
- `cookbook-guest.js` is 1.06 MB raw, 277 kB gzipped.

![image](https://user-images.githubusercontent.com/8503756/180922111-31f99808-632c-4bf0-8ec6-4da2d81f9e56.png)

This small change seems to have inadvertently split some node modules into their own chunks. Now we have:
- `cookbook-main.js` 3.67 MB raw, 872 kB gzipped
- `cookbook-guest.js` 1.06 MB raw, 277 kB gzipped.
- `moment` 711 kB raw, 112.97 kB gzipped
- `RecipeView.js` 87 kB raw, 16 kB gzipped
- `Multiselect.js` 173 kB raw, 45 kB gzipped
- `RecipeList.js` 63 kB raw, 13 kB gzipped
- `RecipeEdit.js` 114 kB raw, 21 kb gzipped
- `NotFound.js` 4 kB raw, 1.9 kB gzipped
- `SearchResult.js` 4.7 kB raw, 1.9 kB gzipped
- `AppIndex.js` 3 kB raw, 1.5 kB gzipped

Apparently, the chunks will be prefetched in Vue 3.0. I was not able to make them prefetch with the webpack magic comment `import(/* webpackPrefetch: true */ '')` for some reason. I am also unsure how to change the long file names. I tried `output: { filename: '' }` in the webpack config and the magic comment `/* webpackChunkName: "" */`, but no change.

I got some good savings by dynamically loading the `moment` locale:
![image](https://user-images.githubusercontent.com/8503756/180926195-7e356c75-2a9e-4aa9-9ff9-1a6eb774901a.png)

`cookbook-main.js` is almost unchanged and only one locale needs to be loaded

I think we may be better off going after npm packages than our own code.Maybe we can lazy load the mardown editor, replace `ActionInput`, etc. I'm not sure why `calendar-js` and `ical` are needed, but both are pretty big. They seem to not be imported anywhere. I only see reference to them in `package-lock.json` as a dependency of `nextcloud/vue`.

Maybe we can also look into tree shaking.

MarcelRobitaille committed:
Implement naive code splitting

Signed-off-by: Marcel Robitaille <mail@marcelrobitaille.me>

MarcelRobitaille committed:
Dynamically load moment locales

Signed-off-by: Marcel Robitaille <mail@marcelrobitaille.me>

MarcelRobitaille committed:
Ignore calendar-js

Signed-off-by: Marcel Robitaille <mail@marcelrobitaille.me>

christianlupus commented:
> Apparently, the chunks will be prefetched in Vue 3.0. I was not able to make them prefetch with the webpack magic comment `import(/* webpackPrefetch: true */ '')` for some reason. I am also unsure how to change the long file names. I tried `output: { filename: '' }` in the webpack config and the magic comment `/* webpackChunkName: "" */`, but no change.

We are not on Vue 3 but Vue 2. Thus the prefetch might not (yet) be implemented.
Why do you want to change the file names? They are [configured in @nextcloud/webpack-vue-config](https://github.com/nextcloud/webpack-vue-config/blob/master/webpack.config.js)

> Maybe we can also look into tree shaking.

I tree shaking not done/present ATM? Moving other (sub-) components (like the Markdown editor/parser) to their own chunk might be a good option.

MarcelRobitaille committed:
Ignore DatetimePicker

Signed-off-by: Marcel Robitaille <mail@marcelrobitaille.me>

MarcelRobitaille commented:
> We are not on Vue 3 but Vue 2. Thus the prefetch might not (yet) be implemented.

I know. It's supposed to be automatic in Vue 3. As far as I understand, you're supposed to be able to manually enable it for any framework with those webpack magic comments. It's a webpack feature, so it should be independent of the Vue version. I am not sure why it's not working.

> Why do you want to change the file names? They are [configured in @nextcloud/webpack-vue-config](https://github.com/nextcloud/webpack-vue-config/blob/master/webpack.config.js)

The auto-generated chunk names are very long, like `cookbook-vendors-node_modules_nextcloud_vue_dist_Components_Multiselect_js.js`. I would prefer to have `cookbook-appindex.js` over `cookbook-src_components_AppIndex.js` for example. I guess it's done to guarantee a unique name, but we know we only have one `AppIndex`.

MarcelRobitaille committed:
Fix dynamically loading moment locale

`en` is the default and cannot be dynamically loaded

Signed-off-by: Marcel Robitaille <mail@marcelrobitaille.me>

MarcelRobitaille committed:
Only load markdown editor where needed (RecipeView)

Signed-off-by: Marcel Robitaille <mail@marcelrobitaille.me>

MarcelRobitaille commented:
I moved `Showdown` to `RecipeView` because that's the only place it's used.

I also implemented my own `ActionInput`. All I did was remove the bloated imports like `DatetimePicker` which imports `@nextcloud/calendar-js` and `ical`, which are both very big. However, now we're getting into a problem of copying GPL stuff. Maybe we can make a PR to dynamically load `DatetimePicker`.

`cookbook-main.js` is down over 2 MB from starting.

Raw:
![image](https://user-images.githubusercontent.com/8503756/181109336-e7269787-728d-4469-9a51-cd9625830f52.png)
Gzipped:
![image](https://user-images.githubusercontent.com/8503756/181109375-7873ad4f-9f66-41da-89d4-a3e9aa258db1.png)
![image](https://user-images.githubusercontent.com/8503756/181109475-e8e69f94-cce2-4e4e-a2dd-b6a57596fd5a.png)

MarcelRobitaille committed:
Implement SimpleActionInput

Signed-off-by: Marcel Robitaille <mail@marcelrobitaille.me>

MarcelRobitaille committed:
Remove webpack IgnorePlugins

These are not needed with SimpleActionInput

Signed-off-by: Marcel Robitaille <mail@marcelrobitaille.me>

MarcelRobitaille committed:
Run prettier

Signed-off-by: Marcel Robitaille <mail@marcelrobitaille.me>

MarcelRobitaille committed:
Import moment from @nextcloud/moment

We don't depend on moment directly, but rather @nextcloud/moment

Signed-off-by: Marcel Robitaille <mail@marcelrobitaille.me>

MarcelRobitaille committed:
Fix eslint lints

Signed-off-by: Marcel Robitaille <mail@marcelrobitaille.me>

MarcelRobitaille committed:
Fix weird formatting of comment before semicolon

The semicolon is needed here because the line below starts with (.
JavaScript will attempt to call the line that now has a semicolon as a
function.

Signed-off-by: Marcel Robitaille <mail@marcelrobitaille.me>

MarcelRobitaille committed:
Run prettier

Signed-off-by: Marcel Robitaille <mail@marcelrobitaille.me>

MarcelRobitaille committed:
Add attribution to original source of SimpleActionInput

Signed-off-by: Marcel Robitaille <mail@marcelrobitaille.me>

MarcelRobitaille committed:
Fix tabs and weird comment formatting

Signed-off-by: Marcel Robitaille <mail@marcelrobitaille.me>

MarcelRobitaille committed:
Run stylelint-fix

Signed-off-by: Marcel Robitaille <mail@marcelrobitaille.me>

MarcelRobitaille commented:
I think this PR is ready for review.

Unfortunately, the way in which the bloat has to be stripped from some libraries requires hacks like [this](https://github.com/nextcloud/cookbook/pull/1116/files#diff-1fb26bc12ac780c7ad7325730ed09fc4c2c3d757c276c3dacc44bfe20faf166f). This makes it possible for one of these changes to cause unexpected breakage. I tried to test as best as I could, but I would appreciate if somebody else could take a look too to double check. I have manually tested:

- `ActionInput` in `AppControls`
- `ActionInput` in `AppNavi`
- Markdown rendering in `RecipeView`: Showdown was moved out of main and only into `RecipeView` where it is used
- `moment`: I changed the language to French and checked that the locale was dynamically loaded and that `moment.locales()` included `fr`
- Going to all pages to check that [loading the components dynamically](https://github.com/nextcloud/cookbook/pull/1116/files#diff-3210ca9ed615e6f0fc1743fc09be3c86bc38178dcc050e5a41ff7982d15875b7) is working properly

Also, I copied and modified `@nextcloud/vue`'s [`ActionInput.vue`](https://github.com/nextcloud/nextcloud-vue/blob/master/src/components/NcActionInput/NcActionInput.vue). I gave the attribution, and we're both AGPL 3.0, so it should be fine, right? I can also try to contribute this upstream, but I'm not sure if it would get merged.

christianlupus commented:
> I think this PR is ready for review.

Man, you ran me over with that one.

> Unfortunately, the way in which the bloat has to be stripped from some libraries requires hacks like [this](https://github.com/nextcloud/cookbook/pull/1116/files#diff-1fb26bc12ac780c7ad7325730ed09fc4c2c3d757c276c3dacc44bfe20faf166f). This makes it possible for one of these changes to cause unexpected breakage. I tried to test as best as I could, but I would appreciate if somebody else could take a look too to double check. 

I feel a bit overwhelmed and also not qualified enough to oversee all things here. Looks like magic to me currently. (Maybe a bit more sleep might help :wink:.) Would you @seyfeb consider having a look at this once? I can click around a bit and see if there is some problem with anything I can see. However, this will be only of limited outcome.

> Also, I copied and modified `@nextcloud/vue`'s [`ActionInput.vue`](https://github.com/nextcloud/nextcloud-vue/blob/master/src/components/NcActionInput/NcActionInput.vue). I gave the attribution, and we're both AGPL 3.0, so it should be fine, right? I can also try to contribute this upstream, but I'm not sure if it would get merged.

The license should be ok if the upstream is GPL. I would find it very important if this were at least offered upstream.

This thing with the source version sounded to me like the best solution but I am completely lost how this would be achieved with `npm`.

seyfeb commented:
> I feel a bit overwhelmed and also not qualified enough to oversee all things here. Looks like magic to me currently. (Maybe a bit more sleep might help 😉.) Would you @seyfeb consider having a look at this once? I can click around a bit and see if there is some problem with anything I can see. However, this will be only of limited outcome.

I can try to have a look!

MarcelRobitaille commented:
Maybe I can summarise the changes:

- Code splitting: Each view has its own smaller bundle now. When going to a page, only the common bundle and that page's bundle are loaded. This reduces the main bundle greatly. The relevant change is [here](https://github.com/nextcloud/cookbook/pull/1116/files#diff-3210ca9ed615e6f0fc1743fc09be3c86bc38178dcc050e5a41ff7982d15875b7).
- Only load VueShowdown in RecipeView. The library is only used on that page, so it can go in that page's bundle instead of the main bundle. The relevant change is [here](https://github.com/nextcloud/cookbook/pull/1116/files#diff-9600fbe491aa41e4177a58b954278861968dc2fd4ab46ffa00bc2fd60ff732f7) and [here](https://github.com/nextcloud/cookbook/pull/1116/files#diff-27653c212e1cfe533e4eb2f7d0d3f89604c9de48a09583b4cbbbcbd08a07da79L45-L50).
- Only load the required `moment` locale. First, the moment locales are stripped from the main bundle. There is no "official" way to do this, so [we have to do it in a kind of hacky way](https://github.com/nextcloud/cookbook/pull/1116/files#diff-1fb26bc12ac780c7ad7325730ed09fc4c2c3d757c276c3dacc44bfe20faf166f). Then, because the main bundle doesn't have any locales, we have to dynamically load the user's locale. This is done [here](https://github.com/nextcloud/cookbook/pull/1116/files#diff-27653c212e1cfe533e4eb2f7d0d3f89604c9de48a09583b4cbbbcbd08a07da79R34-R45).
- Finally, the `ActionInput` component from `@nextcloud/vue` has many features that we do not use. These features have large dependencies like a calendar dropdown. The dependencies are included in the bundle even if those features aren't used. Therefore, I forked that component and stripped out the dependencies and features that we don't use. I called this [`SimpleActionInput`](https://github.com/nextcloud/cookbook/pull/1116/files#diff-d91993202ae7c3735d3411f43f03b3453307bbbff84b8e59aff396570a5d271e). This is used [here](https://github.com/nextcloud/cookbook/pull/1116/files#diff-ecb9a3f6bb832cad84643f8280096bfc051e1747d95e849448f1297187e99b51) and [here](https://github.com/nextcloud/cookbook/pull/1116/files#diff-f6e9f6fe1919b6fb703004217b3c963036231942536a66e9c1ad614efd8e45bd).

christianlupus commented:
This could be die to the fact that you are running (like me) from the `/custom_apps/cookbook` location. Please have a look in the network tab of the den console for failing java script chunks. Are they with the prefix `/apps` or `/custom_apps`? The chunks should be present after billing with npm in the js folder. If yes, the correct url is with the real filesystem path (aka `/custom_apps/cookbook`).

christianlupus commented:
I did not test yet. I think, we might need to add something like
```js
__webpack_public_path__ = linkTo('cookbook', 'js')
```
in the `main.js`. Webpack computes the paths to be `/apps/cookbook` which is valid if the app is installed in `apps`. Otherwiese we need to modify things using the public path.

This was the reason for 48188dc.

seyfeb reviewed (commented):
I did not go through _each_ line, but in general it looks good to me.

One thing that bothers me, is the SimpleActionInput. I (hopefully) does the same thing as Nextcloud's ActionInput, but cannot do more than the way it is currently used. Using more functionality would require reimplementing or copying stuff over from the original component. Drawbacks are: It is a very crowded file, because lots of previously imported stuff (styles and JS) were copied into the single file; it needs to be maintained by ourselves instead of relying on a separately maintained component. 

Afaik, the main reason to use this, was to reduce the bundle size. How much does the change to using the component actually save?

MarcelRobitaille committed:
Remove stale comment "markdown" from main

Signed-off-by: Marcel Robitaille <mail@marcelrobitaille.me>

MarcelRobitaille commented on src/components/SimpleActionInput.vue:
```diff
@@ -0,0 +1,643 @@
+  -->
+
+<template>
+    <li class="action">
```
I forget to be honest. I will add it back.

MarcelRobitaille committed:
Perma-link to ActionInput.vue

Signed-off-by: Marcel Robitaille <mail@marcelrobitaille.me>

MarcelRobitaille committed:
Add back certain functionality of ActionInput.vue

I forgot why I changed this to be honest.

Signed-off-by: Marcel Robitaille <mail@marcelrobitaille.me>

MarcelRobitaille commented on src/main.js:
```diff
@@ -6,7 +6,7 @@
  */
 
 // Markdown
```
Done!

MarcelRobitaille commented on src/components/SimpleActionInput.vue:
```diff
@@ -0,0 +1,643 @@
+  - along with this program. If not, see <http://www.gnu.org/licenses/>.
+  -
+  - This file is modified from:
+  - https://github.com/nextcloud/nextcloud-vue/blob/master/src/components/NcActionInput/NcActionInput.vue
```
That's a very good suggestion. I have done it.

MarcelRobitaille commented:
Thanks for reviewing it @seyfeb 

> Afaik, the main reason to use this, was to reduce the bundle size. How much does the change to using the component actually save?

Correct. The savings is about 750kB.
- With `@nextcloud/vue` `ActionInput` the main.js is 2.25MB
- With my `SimpleActionInput` the main is 1.49MB

The reason is that `@nextcloud/vue` is using multiple large libraries to facilitate the calendar features. I have highlighted some of the big ones in the below bundle analyzer screenshot (I may have missed some of the smaller ones).

![image](https://user-images.githubusercontent.com/8503756/197994304-478aae32-f645-4993-9265-0cda79eaee2f.png)

> One thing that bothers me, is the SimpleActionInput. I (hopefully) does the same thing as Nextcloud's ActionInput, but cannot do more than the way it is currently used. Using more functionality would require reimplementing or copying stuff over from the original component. Drawbacks are: It is a very crowded file, because lots of previously imported stuff (styles and JS) were copied into the single file; it needs to be maintained by ourselves instead of relying on a separately maintained component.

I agree. I don't love this solution either. In my opinion, 33% savings is worth it.

One easy thing that I think actually has a chance of being upstreamed is to use dynamic `import()`s for those libraries only if those features are used. This would have a similar effect for us while not removing functionality. I think it would be a pretty small code change too.

seyfeb commented:
Okay that's actually a lot more than suspected! This might indeed be worth the trouble (you already did the most work anyway). If you see dynamic imports as a better solution, you may change this, but either way I'm fine with the PR

seyfeb commented:
I'm okay with that :)

christianlupus commented:
What did you do, @seyfeb, to get it running? This patch was not yet included, am I wrong?

seyfeb commented:
> What did you do, @seyfeb, to get it running? This patch was not yet included, am I wrong?

Naah, because I hacked it in there to get it running on my local machine. I did override the config of the `publicPath` in `webpack.dev.js` by adding

```
    output: {
        publicPath: path.join('/custom_apps/', appName, '/js/'),
    },
```

christianlupus commented:
> > What did you do, @seyfeb, to get it running? This patch was not yet included, am I wrong?
> 
> Naah, because I hacked it in there to get it running on my local machine. I did override the config of the `publicPath` in `webpack.dev.js` by adding
> 
> ```
>     output: {
>         publicPath: path.join('/custom_apps/', appName, '/js/'),
>     },
> ```

This is not good. It will break any other installation without the `/custom_apps` prefix out there. We need to address this before merging.

christianlupus reviewed (changes requested):
Make the path where the parts are loaded generic and independent of the storage path of the app.

PS: This is just to prevent accidental merging

seyfeb commented:
It's only an issue if you manually add it to custom_apps though? apps installed to apps/ (via the app store) should just work, no?

christianlupus commented:
Unfortunately no. In the [official admin documentation of the server](Https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/config_sample_php_parameters.html#apps) explicitly allows to have different storage locations for apps. I guess that many have the default prefix of apps but this is just a guess. For sure, this will not be working well in the other cases.

seyfeb commented:
I see. I'm a bit confused, but what seems to be working is setting 

```
webpackConfig.output.publicPath = undefined
```
in `webpack.config.js`, hence overriding the definition of the nextcloud default config, which confusingly sets it to `publicPath: path.join('/apps/', appName, '/js/')`

I'm not sure if this makes sense.

christianlupus commented:
I have used the `__webpack_public_path__` approach to test it. For me, I get CSP errors:
![grafik](https://user-images.githubusercontent.com/8202922/198336137-0ab34216-0c78-4315-84ab-f468f4e3f0e7.png)

Can you please check that your browser loads all imports as desired? I am using Firefox on Linux.

seyfeb commented:
With my approach described above I'm seemingly  on the good side of things

<img width="1495" alt="Screenshot 2022-10-27 at 19 53 47" src="https://user-images.githubusercontent.com/7623143/198363367-527fddb4-8801-451e-916b-aee9524cc89d.png">

MarcelRobitaille commented:
I have rebased on NC25 and the savings are still quite large excluding `NcActionInput`. However, some of the other Nextcloud Vue components are rather large. I will look into whether anything can be done.

Without `NcActionInput` (1.6MB)
![image](https://user-images.githubusercontent.com/8503756/199258545-ab935abd-eda5-46d9-ab4d-c0e6d776be82.png)

With `NcActionInput` (2.36MB)
![image](https://user-images.githubusercontent.com/8503756/199258657-da7e17be-74da-4f63-a190-32628abc2172.png)
